File: java_algorithm/zip/kruskal/도시_분할_계획.java
package java_algorithm.zip.kruskal;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.Collections;

public class 도시_분할_계획 {
    final static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static ArrayList<Road> roads;
    static int[] parent;

    static class Road implements Comparable<Road> {
        int a;
        int b;
        int c;

        Road(int a, int b, int c) {
            this.a = a;
            this.b = b;
            this.c = c;
        }

        public int compareTo(Road o) {
            return this.c - o.c;
        }
    }

    public static int findParent(int target) {
        if(parent[target] == target) return target;
        else return parent[target] = findParent(parent[target]);
    }

    public static void union(int a, int b) {
        a = findParent(a);
        b = findParent(b);

        if (a < b) parent[b] = a;
        else parent[a] = b;
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        roads = new ArrayList<>();
        parent = new int[n+1];
        for(int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            roads.add(new Road(a, b, c));
        }

        for(int i = 1; i <= n; i++) {
            parent[i] = i;
        }

        // Arrays.sort(roads);
        Collections.sort(roads);

        int ans = 0;
        int max = 0;
        for(int i = 0; i < roads.size(); i++) {
            Road road = roads.get(i);
            if (findParent(road.a) != findParent(road.b)) {
                max = Math.max(max, road.c);
                ans += road.c;
                union(road.a, road.b);
            }
        }

        System.out.println(ans - max);
    }
    
}

File: kotlin/zip/graph/kruskal/도시_분할_계획.kt
// 30분
// 최소 신장 트리

val br = System.`in`.bufferedReader()
val bw = System.out.bufferedWriter()

lateinit var roadList: ArrayList<IntArray>
lateinit var parent: IntArray
var answer = 0

fun union(target1: Int, target2: Int) {
    val parent1 = parent[target1]
    val parent2 = parent[target2]

    if (parent1 < parent2) {
        parent[parent2] = parent[parent1]
    } else {
        parent[parent1] = parent[parent2]
    }
}

fun findParent(target: Int): Int {
    if (parent[target] == target) return target
    else {
        parent[target] = findParent(parent[target])
        return parent[target]
    }
}

fun divideCity(): Int {
    var lastRoad = 0
    for (road in roadList) {
        if (findParent(road[0]) != findParent(road[1])) {
            lastRoad = road[2]
            answer += road[2]
            union(road[0], road[1])
        }
    }  

    return answer - lastRoad
}

fun main() = with(System.out.bufferedWriter()) {
    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    roadList = ArrayList<IntArray>()
    parent = IntArray(n+1) {it}

    repeat(m) {
        val road = br.readLine().split(" ").map {it.toInt()}.toIntArray()
        roadList.add(road)
    }

    roadList.sortWith(compareBy{it[2]})

    val result = divideCity()
    write("$result")
    close()
}

// PriorityQueue 풀이

import java.util.PriorityQueue

val br = System.`in`.bufferedReader()

fun main() = with(System.out.bufferedWriter()) {
    val (n, m) = br.readLine().split(" ").map {it.toInt()}
    val graph = Array(n+1) {mutableListOf<Pair<Int, Int>>()}
    val visited = BooleanArray(n+1)
    val queue = PriorityQueue<Pair<Int, Int>> {o1, o2 -> o1.second - o2.second}

    repeat(m) {
        val (s, e, c) = br.readLine().split(" ").map {it.toInt()}
        graph[s].add(e to c)
        graph[e].add(s to c)
    }

    var cnt = 0
    var ans = 0
    var max = 0

    queue.add(1 to 0)
    while(queue.isNotEmpty()) {
        val (cur, cost) = queue.poll()

        if(visited[cur]) continue

        visited[cur] = true
        cnt++
        ans += cost

        if(max < cost) 
            max = cost

        if (cnt == n)
            break

        // 컬렉션으로 복사
        graph[cur].filterNot {visited[it.first]}.toCollection(queue)
    }

    write("${ans - max}")
    close()
}